CATL-1032: Ability To Use Case Tags In Webforms

vinuvarshith opened the pull request:
Overview
----------------------------------------
We would like to give the user the ability to add case tags while filling a form.
Users can already use 'Tags' which are for contacts from the 'contact 1' tab in the webform configurations. This PR adds a similar feature in the Cases tab for Case Tags.

vinuvarshith committed:
CATL-1032: Ability To Use Case Tags In Webforms

colemanw commented on includes/wf_crm_webform_postprocess.inc:
```diff
@@ -1301,6 +1301,67 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
+              }
+              $result = wf_civicrm_api('EntityTag', 'delete', $remTagsParams);
+            }
+          }
```
Wow that's a very long block of code being added here just to process tags. Isn't there already a helper function for this? If not there should be.

colemanw replied:
Found it. Look at function `wf_crm_webform_postprocess::addOrRemoveMultivaluedData`

vinuvarshith replied:
I have updated the code to use 
wf_crm_webform_postprocess::addOrRemoveMultivaluedData
This also needed some changes to 'addOrRemoveMultivaluedData' function.

colemanw commented on includes/utils.inc:
```diff
@@ -54,7 +54,7 @@ function wf_crm_field_options($field, $context, $data) {
       $ret = wf_crm_get_privacy_options();
     }
-    elseif ($table === 'other') {
+    elseif ($table === 'other' || $table === 'case') {
```
Looks to me like this will break all other pseudoconstant lookups for cases (as the `elseif ($table === 'case)` condition short-circuits all case fields.

Reading the code, I think we are actually safe to remove this outer conditional wrapping groups and tags, so we can just do `elseif ($field['table'] === 'tag') {` and `elseif ($field['table'] === 'group') {`

vinuvarshith replied:
Actually I have fixed this in my local but didnt update the PR. I have updated this now. Let me know if this is better or can we just remove the outer ($table === 'other') wrapper.

colemanw reviewed (changes requested):
I would like to see more reuse of code for handling tag saving, as it should be basically the same as for contacts.

Also, since activity tags are exactly the same as case tags, I think this PR ought to add support for both, and it should require almost no extra code.

colemanw commented:
Thanks for working on this @vinuvarshith. Let me know if those comments make sense.
Here's a gist for how to add the fields for all 3 types of tags without repeating any code.
https://gist.github.com/colemanw/a6621c4b08017f5d0e509830f2173667

vinuvarshith committed:
CATL-1032: Code Improvements

vinuvarshith commented:
@colemanw 
Thank you for reviewing and for the gist.
I have addressed your comments and updated the code. Pls check and let me know if this is ok.
Thanks

colemanw commented on includes/utils.inc:
```diff
@@ -54,12 +54,12 @@ function wf_crm_field_options($field, $context, $data) {
     }
-    elseif ($table === 'other') {
-      if ($field['table'] === 'tag') {
+    elseif ($table === 'other' || ($table === 'case' && strpos($name, 'tag_') === 0)) {
```
This disallows us from getting activity tags.
Could you please take my previous suggestion and remove the outer `elseif`.

vinuvarshith replied:
Updated as suggested.

colemanw commented:
It looks like you've got activity tags exposed to the webform now but haven't yet added postprocessing for them. This is moving in the right direction though.

vinuvarshith committed:
Add Support For Activity Tags

vinuvarshith commented:
@colemanw I have added support for activity tags too. Please check.
Thanks

colemanw commented on includes/wf_crm_webform_postprocess.inc:
```diff
@@ -1442,6 +1450,33 @@ class wf_crm_webform_postprocess extends wf_crm_webform_base {
+    $removeTags = array_diff($allTags, $tagValues);
+    $removeTags = array_combine($removeTags, $removeTags);
+    $this->addOrRemoveMultivaluedData('tag', $entityType, $entityId, $tagValues, $removeTags);
+  }
```
I'm confused about what's going on with this function.
`addOrRemoveMultivaluedData` is sufficient to handle checkbox-style tagging widgets. It looks like you're trying to support something else - free tagging perhaps? Can you give more details about how this is being used?

jamienovick replied:
@vinuvarshith Can you respond to coleman here please?

vinuvarshith replied:
@colemanw
What I am trying to do here is to support adding and untagging. With 'addOrRemoveMultivaluedData' we pass $removeTags param which should have the tags to be removed.
**What I tried previously:**
So if we try to remove a tag from webform by unchecking the checkbox, that value is not recorded and so is not handled. Here I am getting all tag ids, calculating a diff between that tags to be added and all tags which I consider as tags to be removed.
**Updated logic:**
Now I have updated this code to get the values from $this->crmValues. Is this ok? Should I also get the checked tag values from $this->crmValues too instead of params?

colemanw replied:
Thanks for the clarification.
It still appears to me that you are reinventing the wheel here, and it's a much bigger wheel than necessary. The function `saveGroupsAndTags` calculates the `$add` and `$remove` values with 2 lines of code. Why not do it that way?

See https://github.com/colemanw/webform_civicrm/blob/b54ddfd6374a80089bbe03b12be55ac85c20786b/includes/wf_crm_webform_postprocess.inc#L860-L873

vinuvarshith replied:
Thanks for pointing this out. I have checked and fixed 'handleEntityTags' based on 'saveGroupsAndTags'. Tested in a clean setup and this works fine for me.
@colemanw

colemanw committed:
Fix contact tagsets

colemanw reviewed (commented):
This PR is really looking good now! The code is nice & clean and the support for case/activity tags is a good feature.

However I'm not sure what exactly the `handleEntityTags` function is doing. Can you clarify?

colemanw commented:
@vinuvarshith this is waiting on your response.

colemanw commented:
@vinuvarshith ping

jamienovick commented:
@vinuvarshith Can you respond to coleman here please?

vinuvarshith committed:
CATL-1032: Update 'handleEntityTags' Method Get Values From 'crmValues'

vinuvarshith commented:
@colemanw 
I am so sorry. I somehow missed the notifications and completely forgot about this PR.
I have updated and commented about 'handleEntityTags'. Pls check and let me know if I can make it better.
Thanks for your time in reviewing this PR. 

Apologies once again

colemanw reviewed (changes requested):
I've commented on the `handleEntityTags` function and I think it can be further simplified.
Once that is done I can help you rebase this PR over to the 7.x-5.x branch as we're not going to be adding any new features to the 4.x branch.

vinuvarshith committed:
CATL-1032: Fix And Update 'handleEntityTags' Function And Fix Prefill Of Activity Tags

vinuvarshith commented:
@colemanw I have replied and updated the code based on your comments. Pls let me know how to proceed.

colemanw commented:
@jamienovick I have rebased this into the 5.x branch, and made some additional tweaks to it. Could you please have someone review/test #268
